src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -2,273 +2,374 @@
+            result[AnalyzerConfigDocuments.Checksum] = AnalyzerConfigDocuments;
+        }
 
-            cancellationToken.ThrowIfCancellationRequested();
+        ChecksumCollection.Find(state.ProjectReferences, ProjectReferences, searchingChecksumsLeft, result, cancellationToken);
+        ChecksumCollection.Find(state.MetadataReferences, MetadataReferences, searchingChecksumsLeft, result, cancellationToken);
+        ChecksumCollection.Find(state.AnalyzerReferences, AnalyzerReferences, searchingChecksumsLeft, result, cancellationToken);
 
-            if (searchingChecksumsLeft.Remove(Checksum))
-            {
-                result[Checksum] = this;
-            }
+        await ChecksumCollection.FindAsync(state.DocumentStates, searchingChecksumsLeft, result, cancellationToken).ConfigureAwait(false);
+        await ChecksumCollection.FindAsync(state.AdditionalDocumentStates, searchingChecksumsLeft, result, cancellationToken).ConfigureAwait(false);
+        await ChecksumCollection.FindAsync(state.AnalyzerConfigDocumentStates, searchingChecksumsLeft, result, cancellationToken).ConfigureAwait(false);
+    }
+}
 
-            if (searchingChecksumsLeft.Remove(Info))
-            {
-                result[Info] = state.Attributes;
-            }
+internal sealed class DocumentStateChecksums(Checksum infoChecksum, Checksum textChecksum) : IChecksummedObject
+{
+    public Checksum Checksum { get; } = Checksum.Create(infoChecksum, textChecksum);
+    public Checksum Info => infoChecksum;
+    public Checksum Text => textChecksum;
 
-            if (searchingChecksumsLeft.Remove(Text))
-            {
-                result[Text] = await SerializableSourceText.FromTextDocumentStateAsync(state, cancellationToken).ConfigureAwait(false);
-            }
-        }
+    public void AddAllTo(HashSet<Checksum> checksums)
+    {
+        checksums.Add(this.Checksum);
+        checksums.Add(this.Info);
+        checksums.Add(this.Text);
     }
 
-    /// <summary>
-    /// hold onto object checksum that currently doesn't have a place to hold onto checksum
-    /// </summary>
-    internal static class ChecksumCache
+    public void Serialize(ObjectWriter writer)
     {
-        private static readonly ConditionalWeakTable<object, object> s_cache = new();
+        // Writing this is optional, but helps ensure checksums are being computed properly on both the host and oop side.
+        this.Checksum.WriteTo(writer);
+        this.Info.WriteTo(writer);
+        this.Text.WriteTo(writer);
+    }
 
-        public static IReadOnlyList<T> GetOrCreate<T>(IReadOnlyList<T> unorderedList, ConditionalWeakTable<object, object>.CreateValueCallback orderedListGetter)
-            => (IReadOnlyList<T>)s_cache.GetValue(unorderedList, orderedListGetter);
+    public static DocumentStateChecksums Deserialize(ObjectReader reader)
+    {
+        var checksum = Checksum.ReadFrom(reader);
```

Comment:
>var checksum = Checksum.ReadFrom(reader); [](http://example.com/codeflow?start=8&length=41)

I definitely understand persisting this for solutions and projects, but I'm not sold that it's necessary at the document level. From my understanding, there will be a ton of these (and each DocumentStateChecksums persistence size will increase by about 50%). Additionally, wouldn't the project/solution checksum verification catch this?

Reply:
That seems reasonable. I can potentially do in followup :-)

Reply:
This was the one suggestion I made that I thought was worthwhile changing in this PR. :) Why not do as part of this PR?